matrix-client.el

```diff
@@ -180,113 +199,104 @@ event-handlers and input-filters.")
+If [`matrix-client-use-auth-source'] is non-nil, attempt to log
+in using data from auth-source.  Otherwise, prompt for username
+and password."
+  (let* ((auth-source-creation-prompts (a-list 'username "Matrix identity: "
+                                               'secret "Matrix password for %u (homeserver: %h): "))
          (found (nth 0 (auth-source-search :max 1
                                            :host (oref con :base-url)
                                            :user username
                                            :require '(:user :secret)
                                            :create t))))
-    (when (and
-           found
-           (matrix-login-with-password con
-                                       (plist-get found :user)
-                                       (let ((secret (plist-get found :secret)))
-                                         (if (functionp secret)
-                                             (funcall secret)
-                                           secret)))
-           (oset con :username (plist-get found :user))
-           (let ((save-func (plist-get found :save-function)))
-             (when save-func (funcall save-func)))))))
-
-(defun matrix-client-disconnect (&optional con)
-  "Disconnect from Matrix and kill all active room buffers."
+    (when (and found
+               (matrix-login-with-password con (plist-get found :user)
+                                           (let ((secret (plist-get found :secret)))
+                                             (if (functionp secret)
+                                                 (funcall secret)
+                                               secret))))
+      (oset con :username (plist-get found :user))
+      (when-let ((save-func (plist-get found :save-function)))
+        (funcall save-func)))))
+
+(defun matrix-client-disconnect (&optional connection)
+  "Disconnect from CONNECTION or all Matrix connections, killing room buffers."
   (interactive)
-  (let ((discon (lambda (con)
-                  (dolist (room (oref (cadr con) :rooms))
-                    (kill-buffer (oref (cdr room) :buffer)))
-                  (oset (cadr con) :running nil)
-                  (setq matrix-client-connections
-                        (cl-remove (car con)
-                                   matrix-client-connections
-                                   :test 'equal
-                                   :key 'car)))))
-    (if con
-        (funcall discon con)
-      (dolist (con matrix-client-connections)
-        (funcall discon con)))))
-
-(defmethod matrix-client-start-watchdog ((con matrix-client-connection) &optional force timer-secs)
+  (let ((connections (if connection
+                         (list (cons nil connection))
+                       matrix-client-connections)))
+    (cl-loop for (_ . con) in connections
+             do (progn
+                  ;; TODO: Improve the structure of these lists.  It
```

Comment:
I'd have to agree, this part is very confusing. I don't like the cl loop macro very much myself, but I do like the destructuring that it lets us do, so I think it's fine here.